com.hazelcast.map.impl.querycache.QueryCacheTest.testQueryCache_withLocalListener [HZ-1861]

_master_ (commit fa729d9863694df3ae759259910cd0e496774759)

Failed on Sonar build (Oracle JDK 11): http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar/1295/testReport/com.hazelcast.map.impl.querycache/QueryCacheTest/testQueryCache_withLocalListener/

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: Count of add events wrong! expected:<9> but was:<10>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at com.hazelcast.map.impl.querycache.QueryCacheTest.lambda$testQueryCache_withLocalListener$1(QueryCacheTest.java:125)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1226)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1328)
	at com.hazelcast.map.impl.querycache.QueryCacheTest.testQueryCache_withLocalListener(QueryCacheTest.java:125)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

</details>

<details><summary>Standard output:</summary>

```
22:28:43,597  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTTING_DOWN
22:28:43,597  WARN |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Terminating forcefully...
22:28:43,597  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
22:28:43,598  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], remoteEndpoint=[address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], alive=false}
22:28:43,598  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], remoteEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], alive=false}
22:28:43,598  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], remoteEndpoint=[address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], alive=false}
22:28:43,598  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5703, uuid=e27950da-518b-4b5d-8eb0-611b07b0e5cd], remoteEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], alive=false}
22:28:43,598  WARN |testQueryCache_withLocalListener| - [MembershipManager] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Member [127.0.0.1]:5703 - e27950da-518b-4b5d-8eb0-611b07b0e5cd is suspected to be dead for reason: Connection manager is stopped on Member [127.0.0.1]:5703 - e27950da-518b-4b5d-8eb0-611b07b0e5cd this
22:28:43,598  INFO |testQueryCache_withLocalListener| - [MembershipManager] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Removing Member [127.0.0.1]:5703 - e27950da-518b-4b5d-8eb0-611b07b0e5cd
22:28:43,598  INFO |testQueryCache_withLocalListener| - [PartitionStateManager] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID e27950da-518b-4b5d-8eb0-611b07b0e5cd
22:28:43,598  INFO |testQueryCache_withLocalListener| - [PartitionStateManager] hz.recursing_wozniak.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID e27950da-518b-4b5d-8eb0-611b07b0e5cd
22:28:43,598  INFO |testQueryCache_withLocalListener| - [ClusterService] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] 

Members {size:2, ver:4} [
	Member [127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b this
	Member [127.0.0.1]:5702 - 0ccdad1c-95fe-4f6a-981a-54ffecd605ad
]

22:28:43,598  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Shutting down node engine...
22:28:43,598  INFO |testQueryCache_withLocalListener| - [TransactionManagerService] hz.trusting_wozniak.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: e27950da-518b-4b5d-8eb0-611b07b0e5cd
22:28:43,599  INFO |testQueryCache_withLocalListener| - [ClusterService] hz.recursing_wozniak.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] 

Members {size:2, ver:4} [
	Member [127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b
	Member [127.0.0.1]:5702 - 0ccdad1c-95fe-4f6a-981a-54ffecd605ad this
]

22:28:43,599  INFO |testQueryCache_withLocalListener| - [MigrationManager] hz.trusting_wozniak.migration - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Repartitioning cluster data. Migration tasks count: 1
22:28:43,598  INFO |testQueryCache_withLocalListener| - [TransactionManagerService] hz.recursing_wozniak.cached.thread-5 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: e27950da-518b-4b5d-8eb0-611b07b0e5cd
22:28:43,601  INFO |testQueryCache_withLocalListener| - [NodeExtension] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
22:28:43,601  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 4 ms.
22:28:43,601  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTDOWN
22:28:43,601  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
22:28:43,601  WARN |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Terminating forcefully...
22:28:43,601  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
22:28:43,601  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], remoteEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], alive=false}
22:28:43,602  INFO |testQueryCache_withLocalListener| - [MockServer] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], remoteEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], alive=false}
22:28:43,602  INFO |testQueryCache_withLocalListener| - [MembershipManager] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Removing Member [127.0.0.1]:5702 - 0ccdad1c-95fe-4f6a-981a-54ffecd605ad
22:28:43,602  INFO |testQueryCache_withLocalListener| - [PartitionStateManager] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID 0ccdad1c-95fe-4f6a-981a-54ffecd605ad
22:28:43,602  INFO |testQueryCache_withLocalListener| - [ClusterService] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] 

Members {size:1, ver:5} [
	Member [127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b this
]

22:28:43,602  INFO |testQueryCache_withLocalListener| - [TransactionManagerService] hz.trusting_wozniak.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5702, UUID: 0ccdad1c-95fe-4f6a-981a-54ffecd605ad
22:28:43,602  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Shutting down node engine...
22:28:43,603  INFO |testQueryCache_withLocalListener| - [MigrationRequestOperation] hz.recursing_wozniak.async.thread-4 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Failure while executing MigrationInfo{uuid=3618776c-19c2-4ec5-9f56-67e59f124448, partitionId=0, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b, destinationCurrentReplicaIndex=2, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=4, partitionVersionIncrement=2, status=ACTIVE}
com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!
	at com.hazelcast.internal.partition.operation.BaseMigrationOperation.verifyMaster(BaseMigrationOperation.java:154) ~[classes/:?]
	at com.hazelcast.internal.partition.operation.MigrationRequestOperation.trySendNewFragment(MigrationRequestOperation.java:179) ~[classes/:?]
	at com.hazelcast.internal.partition.operation.MigrationRequestOperation.access$900(MigrationRequestOperation.java:86) ~[classes/:?]
	at com.hazelcast.internal.partition.operation.MigrationRequestOperation$MigrationCallback.lambda$accept$0(MigrationRequestOperation.java:441) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) ~[classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) ~[classes/:?]
22:28:43,603  WARN |testQueryCache_withLocalListener| - [MigrationManager] hz.trusting_wozniak.async.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Failed migration from Member [127.0.0.1]:5702 - 0ccdad1c-95fe-4f6a-981a-54ffecd605ad for MigrationInfo{uuid=3618776c-19c2-4ec5-9f56-67e59f124448, partitionId=0, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b, destinationCurrentReplicaIndex=2, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=4, partitionVersionIncrement=2, status=ACTIVE}
com.hazelcast.core.MemberLeftException: Member [127.0.0.1]:5702 - 0ccdad1c-95fe-4f6a-981a-54ffecd605ad has left cluster!
	at com.hazelcast.spi.impl.operationservice.impl.InvocationMonitor$OnMemberLeftTask.onTargetLoss(InvocationMonitor.java:427) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationMonitor$OnMemberLeftTask.run0(InvocationMonitor.java:394) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationMonitor$MonitorTask.run(InvocationMonitor.java:254) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
22:28:43,602  WARN |testQueryCache_withLocalListener| - [Invocation] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Future.complete(Object) on completed future. Request: Invocation{op=com.hazelcast.internal.partition.operation.MigrationOperation{serviceName='hz:core:partitionService', identityHash=190659734, partitionId=0, replicaIndex=1, callId=-4, invocationTime=1659392923600 (2022-08-01 22:28:43.600), waitTimeout=-1, callTimeout=300000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0, migration=MigrationInfo{uuid=3618776c-19c2-4ec5-9f56-67e59f124448, partitionId=0, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b, destinationCurrentReplicaIndex=2, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=4, partitionVersionIncrement=2, status=ACTIVE}}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=300000, firstInvocationTimeMs=1659392923601, firstInvocationTime='2022-08-01 22:28:43.601', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[127.0.0.1]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=com.hazelcast.internal.server.FirewallingServer$FirewallingConnection@67f99bc1}, current value: Packet{partitionId=-1, frameLength=44, conn=MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], remoteEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], alive=false}, rawFlags=10010011, isUrgent=true, packetType=OPERATION, typeSpecificFlags=[isResponse=true, isOpControl=false]}, offered value: ExceptionalResult{cause=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!}
java.lang.Exception: Future.complete(Object) on completed future. Request: Invocation{op=com.hazelcast.internal.partition.operation.MigrationOperation{serviceName='hz:core:partitionService', identityHash=190659734, partitionId=0, replicaIndex=1, callId=-4, invocationTime=1659392923600 (2022-08-01 22:28:43.600), waitTimeout=-1, callTimeout=300000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0, migration=MigrationInfo{uuid=3618776c-19c2-4ec5-9f56-67e59f124448, partitionId=0, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b, destinationCurrentReplicaIndex=2, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=4, partitionVersionIncrement=2, status=ACTIVE}}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=300000, firstInvocationTimeMs=1659392923601, firstInvocationTime='2022-08-01 22:28:43.601', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[127.0.0.1]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=com.hazelcast.internal.server.FirewallingServer$FirewallingConnection@67f99bc1}, current value: Packet{partitionId=-1, frameLength=44, conn=MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=0ccdad1c-95fe-4f6a-981a-54ffecd605ad], remoteEndpoint=[address=[127.0.0.1]:5701, uuid=f96c88df-91e2-4808-8359-eb89a5fbbe0b], alive=false}, rawFlags=10010011, isUrgent=true, packetType=OPERATION, typeSpecificFlags=[isResponse=true, isOpControl=false]}, offered value: ExceptionalResult{cause=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!}
	at com.hazelcast.spi.impl.AbstractInvocationFuture.warnIfSuspiciousDoubleCompletion(AbstractInvocationFuture.java:1256) ~[classes/:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.complete0(AbstractInvocationFuture.java:1230) ~[classes/:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.completeExceptionallyInternal(AbstractInvocationFuture.java:1223) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.completeExceptionally(Invocation.java:680) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.notifyThrowable(Invocation.java:386) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.notifyError(Invocation.java:330) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationRegistry.shutdown(InvocationRegistry.java:229) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.shutdownInvocations(OperationServiceImpl.java:534) ~[classes/:?]
	at com.hazelcast.spi.impl.NodeEngineImpl.shutdown(NodeEngineImpl.java:543) ~[classes/:?]
	at com.hazelcast.instance.impl.Node.shutdownServices(Node.java:605) ~[classes/:?]
	at com.hazelcast.instance.impl.Node.shutdown(Node.java:541) ~[classes/:?]
	at com.hazelcast.instance.impl.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:101) ~[classes/:?]
	at com.hazelcast.instance.impl.LifecycleServiceImpl.terminate(LifecycleServiceImpl.java:89) ~[classes/:?]
	at com.hazelcast.test.mocknetwork.TestNodeRegistry.shutdown(TestNodeRegistry.java:139) ~[test-classes/:?]
	at com.hazelcast.test.mocknetwork.TestNodeRegistry.terminate(TestNodeRegistry.java:129) ~[test-classes/:?]
	at com.hazelcast.test.TestHazelcastInstanceFactory.terminateAll(TestHazelcastInstanceFactory.java:376) ~[test-classes/:?]
	at com.hazelcast.test.HazelcastTestSupport.shutdownNodeFactory(HazelcastTestSupport.java:183) ~[test-classes/:?]
	at jdk.internal.reflect.GeneratedMethodAccessor47.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:381) ~[test-classes/:?]
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37) ~[test-classes/:?]
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104) ~[test-classes/:?]
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:63) ~[test-classes/:?]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) ~[junit-4.13.2.jar:4.13.2]
	at com.hazelcast.test.HazelcastParallelClassRunner.access$301(HazelcastParallelClassRunner.java:49) ~[test-classes/:?]
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162) ~[test-classes/:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
22:28:43,603  WARN |testQueryCache_withLocalListener| - [MigrationManager] hz.trusting_wozniak.async.thread-2 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Migration failed: MigrationInfo{uuid=3618776c-19c2-4ec5-9f56-67e59f124448, partitionId=0, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5701 - f96c88df-91e2-4808-8359-eb89a5fbbe0b, destinationCurrentReplicaIndex=2, destinationNewReplicaIndex=1, master=[127.0.0.1]:5701, initialPartitionVersion=4, partitionVersionIncrement=2, status=ACTIVE}
22:28:43,604  INFO |testQueryCache_withLocalListener| - [MigrationManager] hz.trusting_wozniak.migration - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Rebalance process was failed. Ignoring remaining migrations. Will recalculate the new migration plan. (repartitionTime=Mon Aug 01 22:28:43 UTC 2022, plannedMigrations=1, completedMigrations=1, remainingMigrations=0, totalCompletedMigrations=1, elapsedMigrationOperationTime=3ms, totalElapsedMigrationOperationTime=3ms, elapsedDestinationCommitTime=0ms, totalElapsedDestinationCommitTime=0ms, elapsedMigrationTime=4ms, totalElapsedMigrationTime=4ms)
22:28:43,608  INFO |testQueryCache_withLocalListener| - [NodeExtension] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
22:28:43,608  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 7 ms.
22:28:43,608  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
22:28:43,608  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
22:28:43,608  WARN |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Terminating forcefully...
22:28:43,608  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
22:28:43,608  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Shutting down node engine...
22:28:43,611  INFO |testQueryCache_withLocalListener| - [NodeExtension] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
22:28:43,611  INFO |testQueryCache_withLocalListener| - [Node] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 3 ms.
22:28:43,611  INFO |testQueryCache_withLocalListener| - [LifecycleService] Thread-11047 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after testQueryCache_withLocalListener(com.hazelcast.map.impl.querycache.QueryCacheTest): BuildInfo{version='5.2-SNAPSHOT', build='20220801', buildNumber=20220801, revision=fa729d9, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testQueryCache_withLocalListener(com.hazelcast.map.impl.querycache.QueryCacheTest):'
22:26:40, accumulated pauses: 4980 ms, max pause: 16 ms, pauses over 1000 ms: 0
22:26:45, accumulated pauses: 4983 ms, max pause: 14 ms, pauses over 1000 ms: 0
22:26:50, accumulated pauses: 4986 ms, max pause: 11 ms, pauses over 1000 ms: 0
22:26:55, accumulated pauses: 4978 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:00, accumulated pauses: 4989 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:05, accumulated pauses: 4978 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:10, accumulated pauses: 4977 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:15, accumulated pauses: 4987 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:20, accumulated pauses: 4977 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:25, accumulated pauses: 4990 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:30, accumulated pauses: 4974 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:35, accumulated pauses: 4985 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:40, accumulated pauses: 4986 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:45, accumulated pauses: 4976 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:50, accumulated pauses: 4987 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:27:55, accumulated pauses: 4980 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:00, accumulated pauses: 4983 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:05, accumulated pauses: 4984 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:10, accumulated pauses: 4976 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:15, accumulated pauses: 4981 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:20, accumulated pauses: 4986 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:25, accumulated pauses: 4986 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:30, accumulated pauses: 4975 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:35, accumulated pauses: 4984 ms, max pause: 10 ms, pauses over 1000 ms: 0
22:28:40, accumulated pauses: 3588 ms, max pause: 436 ms, pauses over 1000 ms: 0
```

</details>

Reopening as issue occurred again but on: _5.2.z_ (commit b8a5f14b7e12b6ed807aab4914c7f5be8c58d170)
with oracle-11: https://jenkins.hazelcast.com/job/Hazelcast-5.maintenance-sonar/414/testReport/junit/com.hazelcast.map.impl.querycache/QueryCacheTest/testQueryCache_withLocalListener/
<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: Count of add events wrong! expected:<9> but was:<10>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at com.hazelcast.map.impl.querycache.QueryCacheTest.lambda$testQueryCache_withLocalListener$1(QueryCacheTest.java:125)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1247)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1349)
	at com.hazelcast.map.impl.querycache.QueryCacheTest.testQueryCache_withLocalListener(QueryCacheTest.java:125)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)

```
</details>



Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-5.maintenance-sonar/414/

@burakgok could you please backport your fix to 5.2.z ?

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/querycache/QueryCacheTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/querycache/QueryCacheTest.java
@@ -40,8 +40,10 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -102,18 +104,23 @@ public class QueryCacheTest extends AbstractQueryCacheTestSupport {
         for (int i = 0; i < 30; i++) {
             map.put(i, i);
         }
+        final Set<Object> keys = Collections.newSetFromMap(new ConcurrentHashMap<>());
         final AtomicInteger countAddEvent = new AtomicInteger();
         final AtomicInteger countRemoveEvent = new AtomicInteger();
 
         final QueryCache<Integer, Integer> queryCache = map.getQueryCache(cacheName, new EntryAdapter() {
             @Override
             public void entryAdded(EntryEvent event) {
-                countAddEvent.incrementAndGet();
+                if (keys.add(event.getKey())) {
+                    countAddEvent.incrementAndGet();
+                }
             }
 
             @Override
             public void entryRemoved(EntryEvent event) {
-                countRemoveEvent.incrementAndGet();
+                if (keys.remove(event.getKey())) {
+                    countRemoveEvent.incrementAndGet();
+                }
             }
         }, SQL_PREDICATE, true);
 

